File: design_pattern/kotlin/AbstractFactory.kt
package design_pattern.kotlin

/**
 * [Source](https://refactoring.guru/design-patterns/abstract-factory)
 */


abstract class Button

class WinButton : Button()

class MacButton : Button()

abstract class CheckBox

class WinCheckBox : CheckBox()

class MacCheckBox : CheckBox()

interface GUIFactory {
    fun createButton(): Button

    fun createCheckBox(): CheckBox
}

class WinFactory : GUIFactory {
    override fun createButton(): Button {
        return WinButton()
    }

    override fun createCheckBox(): CheckBox {
        return WinCheckBox()
    }
}

class MacFactory : GUIFactory {
    override fun createButton(): Button {
        return MacButton()
    }

    override fun createCheckBox(): CheckBox {
        return MacCheckBox()
    }
}

fun main() {
    val configOS = "MAC"
    val factory: GUIFactory = if (configOS == "MAC") MacFactory() else WinFactory()
    val button = factory.createButton()
    val checkBox = factory.createCheckBox()
}

File: design_pattern/dart/abstract_factory.dart
abstract class Button {}

class WinButton extends Button {}

class MacButton extends Button {}

abstract class CheckBox {}

class WinCheckBox extends CheckBox {}

class MacCheckBox extends CheckBox {}

abstract class GUIFactory {
  Button createButton();

  CheckBox createCheckBox();
}

class WinFactory implements GUIFactory {
  @override
  Button createButton() => WinButton();

  @override
  CheckBox createCheckBox() => WinCheckBox();
}

class MacFactory implements GUIFactory {
  @override
  Button createButton() => MacButton();

  @override
  CheckBox createCheckBox() => MacCheckBox();
}

void main() {
  final configOS = 'MAC';
  final factory = configOS == 'MAC' ? MacFactory() : WinFactory();
  final button = factory.createButton();
  final checkBox = factory.createCheckBox();
}